File: leetcode/count-all-possible-routes.py
# 1575. Count All Possible Routes
# 🔴 Hard
#
# https://leetcode.com/problems/count-all-possible-routes/
#
# Tags: Array - Dynamic Programming - Memoization

import timeit
from functools import cache
from typing import List


# We can use memoization, starting at each city, we can decide to visit
# any other city, including ones that we have visited previously, any
# time we reach the destination, we could stop there, and that would be
# one of the possible routes, but we can also keep going. If we run out
# of fuel, we return 0 to represent that the branch did not lead to a
# route, even though it could have included some possible routes if it
# went through the finish location at some middle point.
#
# Time complexity: O(m*n) - Where m is the number of locations and n is
# the number of different values that the remaining fuel could take, and
# could be equal to the initial fuel. Once we have computed m*n calls to
# dfs, any possible result will be stored in the cache and calls would
# be solved in O(1)
# Space complexity: O(m*n) - The size of the cache.
#
# Runtime 2055 ms Beats 63.16%
# Memory 33.9 MB Beats 31.58%
class Solution:
    def countRoutes(
        self, locations: List[int], start: int, finish: int, fuel: int
    ) -> int:
        @cache
        def dfs(l: int, fuel: int) -> int:
            if fuel < 0:
                return 0
            return (
                sum(
                    dfs(i, fuel - abs(locations[i] - locations[l]))
                    for i in range(len(locations))
                    if i != l
                )
                + (1 if l == finish else 0)
            ) % 1_000_000_007

        return dfs(start, fuel)


def test():
    executors = [Solution]
    tests = [
        [[4, 3, 1], 1, 0, 6, 5],
        [[5, 2, 1], 0, 2, 3, 0],
        [[2, 3, 6, 8, 4], 1, 3, 5, 4],
    ]
    for executor in executors:
        start = timeit.default_timer()
        for _ in range(1):
            for col, t in enumerate(tests):
                sol = executor()
                result = sol.countRoutes(t[0], t[1], t[2], t[3])
                exp = t[4]
                assert result == exp, (
                    f"\033[93m» {result} <> {exp}\033[91m for"
                    + f" test {col} using \033[1m{executor.__name__}"
                )
        stop = timeit.default_timer()
        used = str(round(stop - start, 5))
        cols = "{0:20}{1:10}{2:10}"
        res = cols.format(executor.__name__, used, "seconds")
        print(f"\033[92m» {res}\033[0m")


test()

File: leetcode/count-all-possible-routes.rs
// 1575. Count All Possible Routes
// 🔴 Hard
//
// https://leetcode.com/problems/count-all-possible-routes/
//
// Tags: Array - Dynamic Programming - Memoization

struct Solution;
impl Solution {
    /// Use dynamic programming, dp[i][j] indicates the number of routes to i
    /// while having still j remaining fuel.
    ///
    /// Time complexity: O(n^2*m) - Three nested loops, two of them iterating
    /// over all the locations, the outer one over 0..fuel+1.
    /// Space complexity: O(m*n) - The size of the dp 2D array.
    ///
    /// Runtime 167 ms Beats 33.33%
    /// Memory 2.2 MB Beats 66.67%
    pub fn count_routes(locations: Vec<i32>, start: i32, finish: i32, fuel: i32) -> i32 {
        let n = locations.len();
        let m = fuel as usize + 1;
        let finish = finish as usize;
        let mut dp = vec![vec![0; m]; n];
        for i in 0..m {
            dp[finish][i] = 1;
            for j in 0..n {
                for k in 0..n {
                    if k == j {
                        continue;
                    }
                    let diff = (locations[j] - locations[k]).abs() as usize;
                    if diff <= i {
                        dp[j][i] += dp[k][i - diff];
                        dp[j][i] %= 1000000007;
                    }
                }
            }
        }
        dp[start as usize][fuel as usize]
    }
}

// Tests.
fn main() {
    let tests = [
        (vec![4, 3, 1], 1, 0, 6, 5),
        (vec![5, 2, 1], 0, 2, 3, 0),
        (vec![2, 3, 6, 8, 4], 1, 3, 5, 4),
    ];
    for t in tests {
        assert_eq!(Solution::count_routes(t.0, t.1, t.2, t.3), t.4);
    }
    println!("\x1b[92m» All tests passed!\x1b[0m")
}
